firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+        W = next(v for v in V if bool(v.ufl_element().value_shape()))
+        if W.ufl_element().family() in ["Raviart-Thomas", "RTCF"]:
+            tdegree = W.ufl_element().degree() - 1
+
+        else:
+            tdegree = W.ufl_element().degree()
+
+        # NOTE: Once extruded is ready, we will need to be aware of this
+        # and construct the appropriate trace space for the HDiv element
+        TraceSpace = FunctionSpace(V.mesh(), "HDiv Trace", tdegree)
+
+        # NOTE: For extruded, we will need to add "on_top" and "on_bottom"
+        trace_conditions = [DirichletBC(TraceSpace, Constant(0.0),
+                                        "on_boundary")]
+
+        # Set up the functions for the original, hybridized
+        # and schur complement systems
+        self.broken_solution = Function(V_d)
+        self.broken_rhs = Function(V_d)
+        self.trace_solution = Function(TraceSpace)
+        self.unbroken_solution = Function(V)
+        self.unbroken_rhs = Function(V)
+
+        # Create the symbolic Schur-reduction
+        Atilde = Tensor(new_form)
+        gammar = TestFunction(TraceSpace)
+        n = FacetNormal(V.mesh())
+
+        # Vector trial function will have a non-empty ufl_shape
+        sigma = next(f for f in TrialFunctions(V_d) if bool(f.ufl_shape))
+
+        # NOTE: Once extruded is ready, this will change slightly
+        # to include both horizontal and vertical interior facets
+        K = Tensor(gammar('+') * ufl.dot(sigma, n) * ufl.dS)
+
+        # Assemble the Schur complement operator and right-hand side
+        self.schur_rhs = Function(TraceSpace)
+        self._assemble_Srhs = create_assembly_callable(
+            K * Atilde.inv * self.broken_rhs,
+            tensor=self.schur_rhs,
+            form_compiler_parameters=context.fc_params)
+
+        schur_comp = K * Atilde.inv * K.T
+        self.S = allocate_matrix(schur_comp,
+                                 bcs=trace_conditions,
+                                 form_compiler_parameters=context.fc_params)
+        self._assemble_S = create_assembly_callable(
+            schur_comp,
+            tensor=self.S,
+            bcs=trace_conditions,
+            form_compiler_parameters=context.fc_params)
+
+        self._assemble_S()
+        self.S.force_evaluation()
+        Smat = self.S.petscmat
+
+        # Nullspace for the multiplier problem
+        nullsp = P.getNullSpace()
+        if nullsp.handle != 0:
+            new_vecs = get_trace_nullspace_vecs(K * Atilde.inv, nullsp,
+                                                V, V_d, TraceSpace)
+            tr_nullsp = PETSc.NullSpace().create(vectors=new_vecs,
+                                                 comm=pc.comm)
+            Smat.setNullSpace(tr_nullsp)
+
+        # Set up the KSP for the system of Lagrange multipliers
+        ksp = PETSc.KSP().create(comm=pc.comm)
+        ksp.setOptionsPrefix(prefix + "trace_")
+        ksp.setTolerances(rtol=1e-13)
```

Comment:
Ah, this is a remnant of previous experimentation. There's no strong reason. I will remove this and amend.